contracts/BSCValidatorSet.sol

```diff
@@ -344,6 +352,46 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
     return CODE_OK;
   }
 
+  function swap(address[] memory addresses, uint i, uint j) internal pure {
+    address tmp = addresses[i];
+    addresses[i]=addresses[j];
+    addresses[j]=tmp;
+  }
+
+  function shuffle(address[] memory validators, uint256 epochNumber, uint startIdx, uint offset, uint limit, uint modNumber) internal pure {
+    for (uint i = 0; i<limit; i++) {
+      uint random = uint(keccak256(abi.encodePacked(epochNumber, validators[startIdx+i]))) % modNumber;
```

Comment:
It is best not to give `Validators` the opportunity to change the random factor

Reply:
updated